server/odc-service/src/main/java/com/oceanbase/odc/service/git/GitIntegrationService.java

```diff
@@ -0,0 +1,161 @@
+import com.oceanbase.odc.metadb.git.GitRepoRepository;
+import com.oceanbase.odc.metadb.git.GitRepositoryEntity;
+import com.oceanbase.odc.service.encryption.EncryptionFacade;
+import com.oceanbase.odc.service.git.model.GitProvider;
+import com.oceanbase.odc.service.git.model.GitRepository;
+import com.oceanbase.odc.service.git.model.GitRepositoryMapper;
+import com.oceanbase.odc.service.git.model.QueryGitRepositoryParams;
+import com.oceanbase.odc.service.git.vcs.VcsFacadeMapper;
+import com.oceanbase.odc.service.iam.auth.AuthenticationFacade;
+
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * @author: liuyizhuo.lyz
+ * @date: 2024/7/29
+ */
+@Service
+@Slf4j
+@Validated
+@SkipAuthorize("internal authenticated")
+public class GitIntegrationService {
+
+    @Autowired
+    private GitRepoRepository gitRepoRepository;
+    @Autowired
+    private VcsFacadeMapper vcsFacadeMapper;
+    @Autowired
+    private AuthenticationFacade authenticationFacade;
+    @Autowired
+    private EncryptionFacade encryptionFacade;
+    private GitRepositoryMapper gitRepositoryMapper = GitRepositoryMapper.INSTANCE;
+
+    public Page<GitRepository> list(@NotNull Long projectId, @NotNull QueryGitRepositoryParams params,
+            @NotNull Pageable pageable) {
+        params.setOrganizationId(authenticationFacade.currentOrganizationId());
+        return gitRepoRepository.find(params, pageable).map(this::entityToModel);
+    }
+
+    public List<GitRepository> listCandidates(@NotNull GitProvider provider) {
+        return vcsFacadeMapper.getVcsFacade(provider).listRepositories(provider.getPersonalAccessToken());
+    }
+
+    public GitRepository detail(@NotNull Long id) {
+        return entityToModel(nullSafeGet(id));
+    }
+
+    public List<GitRepository> batchCreate(@NotNull Long projectId, @NotNull List<GitRepository> repositories) {
+        long organizationId = authenticationFacade.currentOrganizationId();
+        long creatorId = authenticationFacade.currentUserId();
+
+        Set<String> reposInMeta = gitRepoRepository.findByOrganizationIdAndProjectId(organizationId, projectId)
+                .stream().map(GitRepositoryEntity::getSshAddress).collect(Collectors.toSet());
+        repositories.forEach(r -> {
+            if (reposInMeta.contains(r.getSshAddress())) {
+                throw new BadRequestException(String.format("repository %s already exists", r.getSshAddress()));
+            }
+        });
+
+        List<GitRepositoryEntity> entities = repositories.stream()
+                .map(r -> {
+                    GitRepositoryEntity entity = modelToEntity(r);
+                    entity.setProjectId(projectId);
+                    entity.setCreatorId(creatorId);
+                    entity.setOrganizationId(organizationId);
+                    return entity;
+                })
+                .collect(Collectors.toList());
+        List<GitRepositoryEntity> saved = gitRepoRepository.saveAll(entities);
```

Comment:
saveAll is not a batch operation
we should figure out another way if there are too many repositories~

Reply:
sure,  used `GitRepoRepository#batchCreate` to replace this .